src/transport/quic.rs

```diff
@@ -0,0 +1,266 @@
+            .parse(self.config.pkcs12_password.as_ref().unwrap())
+            .with_context(|| "Could not decrypt  `tls.pkcs12` using `tls.pkcs12_password`")?;
+
+        let mut chain: Vec<rustls::Certificate> = parsed
+            .chain
+            .unwrap()
+            .into_iter()
+            .map(|cert| rustls::Certificate(cert.to_der().unwrap()))
+            .rev()
+            .collect();
+        chain.insert(
+            0,
+            rustls::Certificate(
+                parsed
+                    .cert
+                    .to_der()
+                    .with_context(|| "Could not encode server cert as PEM")?,
+            ),
+        );
+
+        let key = rustls::PrivateKey(parsed.pkey.private_key_to_der().unwrap());
+
+        let mut server_crypto = rustls::ServerConfig::builder()
+            .with_safe_defaults()
+            .with_no_client_auth()
+            .with_single_cert(chain, key)
+            .with_context(|| "Server keys invalid")?;
+
+        server_crypto.alpn_protocols = ALPN_QUIC_TUNNEL.iter().map(|&x| x.into()).collect();
+
+        let mut server_config = quinn::ServerConfig::with_crypto(Arc::new(server_crypto));
+        Arc::get_mut(&mut server_config.transport)
+            .unwrap()
+            .datagram_receive_buffer_size(Some(65536))
+            .datagram_send_buffer_size(65536)
+            .max_idle_timeout(Some(Duration::from_secs(10).try_into()?));
+
+        server_config.use_retry(true);
+        let socket = UdpSocket::bind(addr).await?.into_std()?;
+        quinn::Endpoint::new(EndpointConfig::default(), Some(server_config), socket)
+            .with_context(|| "Failed to start server")
+            .map(|(e, i)| QuicAcceptor(e, i))
+    }
+
+
+    async fn accept(&self, a: &mut Self::Acceptor) -> Result<(Self::RawStream, SocketAddr)> {
+        if let Some(connecting) = a.1.next().await {
+            let addr = connecting.remote_address();
+            return Ok((connecting, addr));
+        }
+        Err(anyhow!("endpoint closed"))
+    }
+
+    async fn handshake(&self, conn: Self::RawStream) -> Result<Self::Stream> {
+        let mut new_connection = conn.await?; // handshake happens here
+        // Connection is established now. Wait for client to open a bidirectional stream
+        if let Some(bi_stream_result) = new_connection.bi_streams.next().await {
+            Ok(Self::Stream::new(bi_stream_result?, new_connection.connection))
+        } else {
+            anyhow::bail!("connection closed before bi stream could be established")
+        }
+    }
+
+    async fn connect(&self, addr: &str) -> Result<Self::Stream> {
+        let mut endpoint = quinn::Endpoint::client("[::]:0".parse().unwrap())
+            .with_context(|| "could not open client socket")?;
+        let mut config =
+            quinn::ClientConfig::new(Arc::new(self.client_crypto.as_ref().unwrap().clone()));
+        // server times out afte 10 sec, so send keepalive every 5 sec
```

Comment:
`afte` -> `after`

Also it will be better to make `keep_alive_interval` and `max_idle_timeout` use constants defined on top of this file.

Reply:
moved to constants